File: android/app/src/main/java/com/btsplusplus/fowallet/kline/MKlineItemData.kt
package com.btsplusplus.fowallet.kline

import bitshares.Utils
import bitshares.bigDecimalfromAmount
import org.json.JSONObject
import java.math.BigDecimal
import kotlin.math.pow

class MKlineItemData {

    var dataIndex = 0
    var showIndex = 0

    var isRise: Boolean = false
    var isMaxPrice: Boolean = false
    var isMinPrice: Boolean = false
    var isMax24Vol: Boolean = false

    var nPriceOpen: BigDecimal? = null
    var nPriceClose: BigDecimal? = null
    var nPriceHigh: BigDecimal? = null
    var nPriceLow: BigDecimal? = null

    var n24Vol: BigDecimal? = null              //  成交量
    var n24TotalAmount: BigDecimal? = null      //  成交额
    var nAvgPrice: BigDecimal? = null           //  成交均价

    var ma60: BigDecimal? = null                //  分时图需要显示 / for timeline

    var main_index01: BigDecimal? = null        //  主图指标01-03（根据设置对应不同指标，比如ma1-ma3，或者ema1-ema3，或者boll、ub、lb等）
    var main_index02: BigDecimal? = null
    var main_index03: BigDecimal? = null
    var fOffsetMainIndex01: Float = 0f
    var fOffsetMainIndex02: Float = 0f
    var fOffsetMainIndex03: Float = 0f

    var adv_index01: BigDecimal? = null        //  高级指标（MACD等）
    var adv_index02: BigDecimal? = null
    var adv_index03: BigDecimal? = null
    var fOffsetAdvIndex01: Float = 0f
    var fOffsetAdvIndex02: Float = 0f
    var fOffsetAdvIndex03: Float = 0f

    var vol_ma5: BigDecimal? = null
    var vol_ma10: BigDecimal? = null

    var change: BigDecimal? = null
    var change_percent: BigDecimal? = null

    var fOffsetOpen: Float = 0f
    var fOffsetClose: Float = 0f
    var fOffsetHigh: Float = 0f
    var fOffsetLow: Float = 0f

    var fOffset24Vol: Float = 0f

    var fOffsetMA60: Float = 0f

    var fOffsetVolMA5: Float = 0f
    var fOffsetVolMA10: Float = 0f

    var date: Long = 0

    fun reset() {
        showIndex = 0

        isRise = false
        isMaxPrice = false
        isMinPrice = false
        isMax24Vol = false

        nPriceOpen = null
        nPriceClose = null
        nPriceHigh = null
        nPriceLow = null

        n24Vol = null
        n24TotalAmount = null
        nAvgPrice = null

        ma60 = null

        main_index01 = null
        main_index02 = null
        main_index03 = null

        fOffsetMainIndex01 = 0f
        fOffsetMainIndex02 = 0f
        fOffsetMainIndex03 = 0f

        adv_index01 = null
        adv_index02 = null
        adv_index03 = null
        fOffsetAdvIndex01 = 0f
        fOffsetAdvIndex02 = 0f
        fOffsetAdvIndex03 = 0f

        vol_ma5 = null
        vol_ma10 = null

        change = null
        change_percent = null

        fOffsetOpen = 0f
        fOffsetClose = 0f
        fOffsetHigh = 0f
        fOffsetLow = 0f

        fOffset24Vol = 0f

        fOffsetMA60 = 0f

        fOffsetVolMA5 = 0f
        fOffsetVolMA10 = 0f

        date = 0L
    }

    constructor()

    companion object {

        /**
         *  (public) 解析服务器的K线数据，生成对应的Model。
         *
         *  fillto          - 可为nil。
         *  ceilHandler     - 可为nil。
         */
        fun parseData(data: JSONObject,
                      fillto: MKlineItemData?,
                      base_id: String?,
                      base_precision: Int,
                      quote_precision: Int,
                      ceilHandler: Array<Int>?,
                      percentHandler: Array<Int>?
        ): MKlineItemData {

            var _fillto = fillto
            if (_fillto == null) {
                _fillto = MKlineItemData()
            }

            var _ceilHandler = ceilHandler
            var _percentHandler = percentHandler


            //  保留小数位数 向上取整
            if (_ceilHandler == null) {
                _ceilHandler = arrayOf(base_precision, BigDecimal.ROUND_UP)
            }

            //  涨跌幅的百分比 handler，有效数位4位。百分比加2位小数点。
            if (_percentHandler == null) {
                _percentHandler = arrayOf(4, BigDecimal.ROUND_UP)
            }

            val n_base_precision = BigDecimal.valueOf(10.0.pow(base_precision))
            val n_quote_precision = BigDecimal.valueOf(10.0.pow(quote_precision))

            val cell_scale = _ceilHandler[0]
            val cell_rounding = _ceilHandler[1]

            val percent_scale = _percentHandler[0]
            val percent_rounding = _percentHandler[1]

            val open_base = data.getString("open_base")
            val open_quote = data.getString("open_quote")

            val high_base = data.getString("high_base")
            val high_quote = data.getString("high_quote")

            val low_base = data.getString("low_base")
            val low_quote = data.getString("low_quote")

            val close_base = data.getString("close_base")
            val close_quote = data.getString("close_quote")

            val key = data.getJSONObject("key")

            if (key.getString("base") == base_id) {

                //  price = base/quote

                val n_open_base = bigDecimalfromAmount(open_base, n_base_precision)
                val n_open_quote = bigDecimalfromAmount(open_quote, n_quote_precision)

                val n_high_base = bigDecimalfromAmount(high_base, n_base_precision)
                val n_high_quote = bigDecimalfromAmount(high_quote, n_quote_precision)

                val n_low_base = bigDecimalfromAmount(low_base, n_base_precision)
                val n_low_quote = bigDecimalfromAmount(low_quote, n_quote_precision)

                val n_close_base = bigDecimalfromAmount(close_base, n_base_precision)
                val n_close_quote = bigDecimalfromAmount(close_quote, n_quote_precision)

                val n_open_price = n_open_base.divide(n_open_quote, cell_scale, cell_rounding)
                val n_high_price = n_high_base.divide(n_high_quote, cell_scale, cell_rounding)
                val n_low_price = n_low_base.divide(n_low_quote, cell_scale, cell_rounding)
                val n_close_price = n_close_base.divide(n_close_quote, cell_scale, cell_rounding)

                _fillto.n24Vol = bigDecimalfromAmount(data.getString("quote_volume"), n_quote_precision)
                _fillto.n24TotalAmount = bigDecimalfromAmount(data.getString("base_volume"), n_base_precision)
                _fillto.isRise = n_open_price <= n_close_price

                //  REMARK：完全一致、高低也相同
                _fillto.nPriceOpen = n_open_price
                _fillto.nPriceClose = n_close_price
                _fillto.nPriceHigh = n_high_price
                _fillto.nPriceLow = n_low_price
            } else {

                //  price = quote/base

                val n_open_base = bigDecimalfromAmount(open_base, n_quote_precision)
                val n_open_quote = bigDecimalfromAmount(open_quote, n_base_precision)

                val n_high_base = bigDecimalfromAmount(high_base, n_quote_precision)
                val n_high_quote = bigDecimalfromAmount(high_quote, n_base_precision)

                val n_low_base = bigDecimalfromAmount(low_base, n_quote_precision)
                val n_low_quote = bigDecimalfromAmount(low_quote, n_base_precision)

                val n_close_base = bigDecimalfromAmount(close_base, n_quote_precision)
                val n_close_quote = bigDecimalfromAmount(close_quote, n_base_precision)

                val n_open_price = n_open_quote.divide(n_open_base, cell_scale, cell_rounding)
                val n_high_price = n_high_quote.divide(n_high_base, cell_scale, cell_rounding)
                val n_low_price = n_low_quote.divide(n_low_base, cell_scale, cell_rounding)
                val n_close_price = n_close_quote.divide(n_close_base, cell_scale, cell_rounding)

                _fillto.n24Vol = bigDecimalfromAmount(data.getString("base_volume"), n_quote_precision)
                _fillto.n24TotalAmount = bigDecimalfromAmount(data.getString("quote_volume"), n_base_precision)
                _fillto.isRise = n_open_price <= n_close_price

                //  REMARK：开收相同、高低反向
                _fillto.nPriceOpen = n_open_price
                _fillto.nPriceClose = n_close_price
                _fillto.nPriceHigh = n_low_price
                _fillto.nPriceLow = n_high_price
            }

            //  成交均价
            if (_fillto.n24Vol!! > BigDecimal.ZERO) {
                _fillto.nAvgPrice = _fillto.n24TotalAmount!!.divide(_fillto.n24Vol, cell_scale, cell_rounding)
            } else {
                _fillto.nAvgPrice = null
            }

            //  计算涨跌额和涨跌幅
            _fillto.change = _fillto.nPriceClose!!.subtract(_fillto.nPriceOpen).setScale(cell_scale, cell_rounding)
            var rate = _fillto.nPriceClose!!.divide(_fillto.nPriceOpen, percent_scale, percent_rounding)
            rate = rate.subtract(BigDecimal.ONE)
            _fillto.change_percent = rate.scaleByPowerOfTen(2).setScale(percent_scale, percent_rounding)

            //  解析日期
            _fillto.date = Utils.parseBitsharesTimeString(key.getString("open"))

            return _fillto
        }

    }

}

File: ios/oplayer/Model/MKlineItemData.m
//
//  MKlineItemData.m
//  oplayer
//
//  Created by SYALON on 13-11-20.
//
//

#import "MKlineItemData.h"
#import "OrgUtils.h"

@interface MKlineItemData()
{
}

@end

@implementation MKlineItemData

@synthesize dataIndex;
@synthesize showIndex;
@synthesize isRise, isMaxPrice, isMinPrice, isMax24Vol;
@synthesize nPriceOpen, nPriceClose, nPriceHigh, nPriceLow;
@synthesize n24Vol, n24TotalAmount, nAvgPrice;
@synthesize ma60;
@synthesize main_index01, main_index02, main_index03, fOffsetMainIndex01, fOffsetMainIndex02, fOffsetMainIndex03;
@synthesize adv_index01, adv_index02, adv_index03, fOffsetAdvIndex01, fOffsetAdvIndex02, fOffsetAdvIndex03;
@synthesize vol_ma5, vol_ma10;
@synthesize change, change_percent;
@synthesize fOffsetOpen, fOffsetClose, fOffsetHigh, fOffsetLow;
@synthesize fOffset24Vol;
@synthesize fOffsetMA60;
@synthesize fOffsetVolMA5, fOffsetVolMA10;
@synthesize date;

- (void)dealloc
{
}

- (void)reset
{
    self.dataIndex = 0;
    self.showIndex = 0;
    
    self.isRise = NO;
    self.isMaxPrice = NO;
    self.isMinPrice = NO;
    self.isMax24Vol = NO;
    
    self.nPriceOpen = nil;
    self.nPriceClose = nil;
    self.nPriceHigh = nil;
    self.nPriceLow = nil;
    
    self.n24Vol = nil;
    self.n24TotalAmount = nil;
    self.nAvgPrice = nil;
    
    self.ma60 = nil;
    
    self.main_index01 = nil;
    self.main_index02 = nil;
    self.main_index03 = nil;
    self.fOffsetMainIndex01 = 0;
    self.fOffsetMainIndex02 = 0;
    self.fOffsetMainIndex03 = 0;
    
    self.adv_index01 = nil;
    self.adv_index02 = nil;
    self.adv_index03 = nil;
    self.fOffsetAdvIndex01 = 0;
    self.fOffsetAdvIndex02 = 0;
    self.fOffsetAdvIndex03 = 0;
    
    self.vol_ma5 = nil;
    self.vol_ma10 = nil;
    
    self.change = nil;
    self.change_percent = nil;
    
    self.fOffsetOpen = 0;
    self.fOffsetClose = 0;
    self.fOffsetHigh = 0;
    self.fOffsetLow = 0;
    
    self.fOffset24Vol = 0;
    
    self.fOffsetMA60 = 0;
    
    self.fOffsetVolMA5 = 0;
    self.fOffsetVolMA10 = 0;
    
    self.date = 0;
}

/**
 *  (public) 解析服务器的K线数据，生成对应的Model。
 *
 *  fillto          - 可为nil。
 *  ceilHandler     - 可为nil。
 */
+ (MKlineItemData*)parseData:(id)data
                      fillto:(MKlineItemData*)fillto
                     base_id:(NSString*)base_id
              base_precision:(NSInteger)base_precision
             quote_precision:(NSInteger)quote_precision
                 ceilHandler:(NSDecimalNumberHandler*)ceilHandler
              percentHandler:(NSDecimalNumberHandler*)percentHandler
{
    if (!fillto){
        fillto = [[MKlineItemData alloc] init];
    }
    
    //  保留小数位数 向上取整
    if (!ceilHandler){
        ceilHandler = [NSDecimalNumberHandler decimalNumberHandlerWithRoundingMode:NSRoundUp
                                                                             scale:base_precision
                                                                  raiseOnExactness:NO
                                                                   raiseOnOverflow:NO
                                                                  raiseOnUnderflow:NO
                                                               raiseOnDivideByZero:NO];
    }
    
    //  涨跌幅的百分比 handler，有效数位4位。百分比加2位小数点。
    if (!percentHandler){
        percentHandler = [NSDecimalNumberHandler decimalNumberHandlerWithRoundingMode:NSRoundUp
                                                                                scale:4
                                                                     raiseOnExactness:NO
                                                                      raiseOnOverflow:NO
                                                                     raiseOnUnderflow:NO
                                                                  raiseOnDivideByZero:NO];
    }

    id open_base = [data objectForKey:@"open_base"];
    id open_quote = [data objectForKey:@"open_quote"];
    
    id high_base = [data objectForKey:@"high_base"];
    id high_quote = [data objectForKey:@"high_quote"];
    
    id low_base = [data objectForKey:@"low_base"];
    id low_quote = [data objectForKey:@"low_quote"];
    
    id close_base = [data objectForKey:@"close_base"];
    id close_quote = [data objectForKey:@"close_quote"];
    
    id key = [data objectForKey:@"key"];
    if ([[key objectForKey:@"base"] isEqualToString:base_id]){
        //  price = base/quote
        id n_open_base = [NSDecimalNumber decimalNumberWithMantissa:[open_base unsignedLongLongValue] exponent:-base_precision isNegative:NO];
        id n_open_quote = [NSDecimalNumber decimalNumberWithMantissa:[open_quote unsignedLongLongValue] exponent:-quote_precision isNegative:NO];
        
        id n_high_base = [NSDecimalNumber decimalNumberWithMantissa:[high_base unsignedLongLongValue] exponent:-base_precision isNegative:NO];
        id n_high_quote = [NSDecimalNumber decimalNumberWithMantissa:[high_quote unsignedLongLongValue] exponent:-quote_precision isNegative:NO];
        
        id n_low_base = [NSDecimalNumber decimalNumberWithMantissa:[low_base unsignedLongLongValue] exponent:-base_precision isNegative:NO];
        id n_low_quote = [NSDecimalNumber decimalNumberWithMantissa:[low_quote unsignedLongLongValue] exponent:-quote_precision isNegative:NO];
        
        id n_close_base = [NSDecimalNumber decimalNumberWithMantissa:[close_base unsignedLongLongValue] exponent:-base_precision isNegative:NO];
        id n_close_quote = [NSDecimalNumber decimalNumberWithMantissa:[close_quote unsignedLongLongValue] exponent:-quote_precision isNegative:NO];
        
        id n_open_price = [n_open_base decimalNumberByDividingBy:n_open_quote withBehavior:ceilHandler];
        id n_high_price = [n_high_base decimalNumberByDividingBy:n_high_quote withBehavior:ceilHandler];
        id n_low_price = [n_low_base decimalNumberByDividingBy:n_low_quote withBehavior:ceilHandler];
        id n_close_price = [n_close_base decimalNumberByDividingBy:n_close_quote withBehavior:ceilHandler];
        
        fillto.n24Vol = [NSDecimalNumber decimalNumberWithMantissa:[[data objectForKey:@"quote_volume"] unsignedLongLongValue]
                                                          exponent:-quote_precision isNegative:NO];
        fillto.n24TotalAmount = [NSDecimalNumber decimalNumberWithMantissa:[[data objectForKey:@"base_volume"] unsignedLongLongValue]
                                                                  exponent:-base_precision isNegative:NO];
        
        //  n_open_price <= n_close_price
        fillto.isRise = [n_open_price compare:n_close_price] != NSOrderedDescending;
        
        //  REMARK：完全一致、高低也相同
        fillto.nPriceOpen = n_open_price;
        fillto.nPriceClose = n_close_price;
        fillto.nPriceHigh = n_high_price;
        fillto.nPriceLow = n_low_price;
    }else{
        //  price = quote/base
        id n_open_base = [NSDecimalNumber decimalNumberWithMantissa:[open_base unsignedLongLongValue] exponent:-quote_precision isNegative:NO];
        id n_open_quote = [NSDecimalNumber decimalNumberWithMantissa:[open_quote unsignedLongLongValue] exponent:-base_precision isNegative:NO];
        
        id n_high_base = [NSDecimalNumber decimalNumberWithMantissa:[high_base unsignedLongLongValue] exponent:-quote_precision isNegative:NO];
        id n_high_quote = [NSDecimalNumber decimalNumberWithMantissa:[high_quote unsignedLongLongValue] exponent:-base_precision isNegative:NO];
        
        id n_low_base = [NSDecimalNumber decimalNumberWithMantissa:[low_base unsignedLongLongValue] exponent:-quote_precision isNegative:NO];
        id n_low_quote = [NSDecimalNumber decimalNumberWithMantissa:[low_quote unsignedLongLongValue] exponent:-base_precision isNegative:NO];
        
        id n_close_base = [NSDecimalNumber decimalNumberWithMantissa:[close_base unsignedLongLongValue] exponent:-quote_precision isNegative:NO];
        id n_close_quote = [NSDecimalNumber decimalNumberWithMantissa:[close_quote unsignedLongLongValue] exponent:-base_precision isNegative:NO];
        
        id n_open_price = [n_open_quote decimalNumberByDividingBy:n_open_base withBehavior:ceilHandler];
        id n_high_price = [n_high_quote decimalNumberByDividingBy:n_high_base withBehavior:ceilHandler];
        id n_low_price = [n_low_quote decimalNumberByDividingBy:n_low_base withBehavior:ceilHandler];
        id n_close_price = [n_close_quote decimalNumberByDividingBy:n_close_base withBehavior:ceilHandler];
        
        fillto.n24Vol = [NSDecimalNumber decimalNumberWithMantissa:[[data objectForKey:@"base_volume"] unsignedLongLongValue]
                                                          exponent:-quote_precision isNegative:NO];
        fillto.n24TotalAmount = [NSDecimalNumber decimalNumberWithMantissa:[[data objectForKey:@"quote_volume"] unsignedLongLongValue]
                                                                  exponent:-base_precision isNegative:NO];
        
        //  n_open_price <= n_close_price
        fillto.isRise = [n_open_price compare:n_close_price] != NSOrderedDescending;
        
        //  REMARK：开收相同、高低反向
        fillto.nPriceOpen = n_open_price;
        fillto.nPriceClose = n_close_price;
        fillto.nPriceHigh = n_low_price;
        fillto.nPriceLow = n_high_price;
    }
    
    //  成交均价
    if ([fillto.n24Vol compare:[NSDecimalNumber zero]] > 0){
        fillto.nAvgPrice = [fillto.n24TotalAmount decimalNumberByDividingBy:fillto.n24Vol withBehavior:ceilHandler];
    }else{
        fillto.nAvgPrice = nil;
    }
    
    //  计算涨跌额和涨跌幅
    fillto.change = [fillto.nPriceClose decimalNumberBySubtracting:fillto.nPriceOpen withBehavior:ceilHandler];
    NSDecimalNumber* rate = [fillto.nPriceClose decimalNumberByDividingBy:fillto.nPriceOpen withBehavior:percentHandler];
    rate = [rate decimalNumberBySubtracting:[NSDecimalNumber one] withBehavior:percentHandler];
    fillto.change_percent = [rate decimalNumberByMultiplyingByPowerOf10:2 withBehavior:percentHandler];
    
    //  解析日期
    fillto.date = [OrgUtils parseBitsharesTimeString:[key objectForKey:@"open"]];
    
    return fillto;
}

@end
